orttraining/orttraining/training_api/module.cc

```diff
@@ -2,12 +2,137 @@
+  Tensor* p_tensor = gradient_.GetMutable<Tensor>();
+  const auto& device = p_tensor->Location().device;
+  if (device.Type() == OrtDevice::CPU) {
+    memset(p_tensor->MutableDataRaw(), 0, p_tensor->SizeInBytes());
+  }
+#if defined(USE_CUDA) || defined(USE_ROCM)
+  else if (device.Type() == OrtDevice::GPU) {
+    ORT_NOT_IMPLEMENTED("Not implemented.");
+  }
+#endif
+  else {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "Unknown device type ", device.Type(), " for param:", name_);
+  }
+  return Status::OK();
+}
+Module::Module(const std::string& train_model_path_or_bytes,
+               std::unordered_map<std::string, std::shared_ptr<Parameter>>& parameters,
+               const std::optional<std::string>& eval_model_path_or_bytes) {
+  // create value copy as same params are passed to Optimizer constructor
+  parameters_ = parameters;
+
+  auto so = onnxruntime::SessionOptions();
+  std::unique_ptr<Environment> env;
+  ORT_THROW_IF_ERROR(Environment::Create(nullptr, env));
+  train_sess_ = std::make_unique<onnxruntime::InferenceSession>(so, *env);
+  ORT_THROW_IF_ERROR(train_sess_->Load(train_model_path_or_bytes));
+  ORT_THROW_IF_ERROR(train_sess_->Initialize());
+  if (eval_model_path_or_bytes.has_value()) {
+    std::shared_ptr<onnxruntime::Model> eval_model;
+    ORT_THROW_IF_ERROR(onnxruntime::Model::Load(eval_model_path_or_bytes.value(), eval_model,
+                                                nullptr, env->GetLoggingManager()->DefaultLogger()));
+    GetGraphInputOutputNames(eval_model->MainGraph(), eval_input_names_, eval_output_names_);
+    // TODO:: do validation on eval inputs and outputs: eg order of user inputs, weights
+    eval_sess_ = std::make_unique<onnxruntime::InferenceSession>(so, *env);
+    ORT_THROW_IF_ERROR(eval_sess_->Load(eval_model_path_or_bytes.value()));
+    ORT_THROW_IF_ERROR(eval_sess_->Initialize());
+  }
+  auto& train_sess_state = train_sess_->GetSessionState();
+
+  std::shared_ptr<onnxruntime::Model> train_model;
+  ORT_THROW_IF_ERROR(onnxruntime::Model::Load(train_model_path_or_bytes, train_model,
+                                              nullptr, env->GetLoggingManager()->DefaultLogger()));
+  GetGraphInputOutputNames(train_model->MainGraph(), train_input_names_, train_output_names_);
+
+  std::vector<std::string> param_input_names, grad_input_names, user_input_names;
+  std::string param_name;
+  for (auto input_name : train_input_names_) {
+    auto it = parameters_.find(input_name);
+    if (it != parameters_.end()) {
+      param_input_names.emplace_back(input_name);
+      weights_.push_back(it->second->Data());
+    } else if (GetParamNameFromGradient(input_name, param_name)) {
+      grad_input_names.emplace_back(input_name);
+      // create gradient buffer
+      // assert param_name is valid.
+      auto it = parameters_.find(param_name);
+      if (it != parameters_.end()) {
+        ORT_THROW_IF_ERROR(it->second->AllocateGrad(input_name, train_sess_state));
```

Comment:
When the param.gradient is in initial state (e.g. without gradient accumulated yet), I think it does not need to be a full sized tensor, in our case , a full size 0 tensor. This will occupy the memory during all times of the training, which it does need to be.  I would suggest to use techniques like strides tensors to describe the initial stage, in InplaceAccumulator op, when we first do the accumulation, we replace that with a full size gradients.

Reply:
Will add a TODO here, we can do it as a next step. also added a work item